pkg/operator/controller/externaldns/status.go

```diff
@@ -0,0 +1,296 @@
+				Reason:  "PodsNotScheduled",
+				Message: message,
+			}
+		}
+		return metav1.Condition{
+			Type:    ExternalDNSPodsScheduledConditionType,
+			Status:  metav1.ConditionTrue,
+			Reason:  "AllPodsScheduled",
+			Message: "All pods are scheduled",
+		}
+	}
+
+}
+
+// MergeConditions updates the conditions list with a new condition
+// the condition is added if no conditions of the same type exists
+// otherwise, the condition is merged with the existing condition of the same type
+func mergeConditions(conditions []metav1.Condition, updates ...metav1.Condition) []metav1.Condition {
+	now := metav1.NewTime(clock.Now())
+	for i, update := range updates {
+		add := true
+		for j, cond := range conditions {
+			if cond.Type == update.Type {
+				add = false
+				if conditionChanged(cond, update) {
+					conditions[j] = update
+					conditions[j].LastTransitionTime = now
+					break
+				}
+			}
+		}
+		if add {
+			updates[i].LastTransitionTime = now
+			conditions = append(conditions, updates[i])
+		}
+	}
+	return conditions
+}
+
+func conditionChanged(a, b metav1.Condition) bool {
+	return a.Status != b.Status || a.Reason != b.Reason || a.Message != b.Message
+}
+
+func createDeploymentAvailabilityUnknownCondition() metav1.Condition {
+	return metav1.Condition{
+		Type:               ExternalDNSDeploymentAvailableConditionType,
+		Status:             metav1.ConditionUnknown,
+		Reason:             "DeploymentAvailabilityUnknown",
+		Message:            "The deployment has no Available status condition set",
+		LastTransitionTime: metav1.NewTime(clock.Now()),
+	}
+}
+
+// updateExternalDNSStatus recomputes all conditions given the current deployment and its status
+// and pushes the new externalDNS custom resource with updated status through a call to the client.Update function
+func (r *reconciler) updateExternalDNSStatus(ctx context.Context, externalDNS *operatorv1alpha1.ExternalDNS, currentDeployment *appsv1.Deployment) error {
+	extDNSWithStatus := externalDNS.DeepCopy()
+	if currentDeployment != nil {
+
+		extDNSWithStatus.Status.Conditions = mergeConditions(extDNSWithStatus.Status.Conditions, computeDeploymentAvailableCondition(currentDeployment))
+		extDNSWithStatus.Status.Conditions = mergeConditions(extDNSWithStatus.Status.Conditions, computeMinReplicasCondition(currentDeployment))
+		extDNSWithStatus.Status.Conditions = mergeConditions(extDNSWithStatus.Status.Conditions, computeAllReplicasCondition(currentDeployment))
+
+		extDNSWithStatus.Status.Conditions = mergeConditions(extDNSWithStatus.Status.Conditions, computeDeploymentPodsScheduledCondition(ctx, r.client, currentDeployment))
+	} else {
+		extDNSWithStatus.Status.Conditions = mergeConditions(extDNSWithStatus.Status.Conditions, createDeploymentAvailabilityUnknownCondition())
+	}
+	extDNSWithStatus.Status.ObservedGeneration = extDNSWithStatus.Generation
+	extDNSWithStatus.Status.Zones = extDNSWithStatus.Spec.Zones
+
+	return r.client.Status().Update(ctx, extDNSWithStatus)
```

Comment:
Would be good to add a check of the changed status similar to what cluster-ingress-operator does [here](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/status.go#L71).